src/lib.rs

```diff
@@ -29,21 +29,52 @@ pub enum SocketAddrs {
     Both((SocketAddr, SocketAddr)),
 }
 
-impl From<&(SocketAddr, SocketAddr)> for SocketAddrs {
-    fn from((local, remote): &(SocketAddr, SocketAddr)) -> Self {
-        Self::Both((*local, *remote))
+fn to_socket_addr<T: ToSocketAddrs>(addrs: T) -> Result<SocketAddr, Error> {
+    addrs
+        .to_socket_addrs()?
+        .next()
+        .ok_or_else(|| Error::new(ErrorKind::InvalidInput, "No address found"))
+}
+
+impl<T> TryFrom<&(T, T)> for SocketAddrs
+where
+    T: ToSocketAddrs,
+{
```

Comment:
@mxinden I could use some Rust help here. I _think_ I need to make this something like 
```rust
impl<S, T> TryFrom<&(S, T)> for SocketAddrs
where
    S: ToSocketAddrs,
    T: ToSocketAddrs,
```
so that I can pass tuples with different types in (like `(IpAddr, SocketAddr)`) but I get a Rust error I don't know what to do with:
```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<&(std::option::Option<_>, _)>` for type `SocketAddrs`
  --> src/lib.rs:54:1
   |
39 | / impl<S, T> TryFrom<&(S, T)> for SocketAddrs
40 | | where
41 | |     S: ToSocketAddrs,
42 | |     T: ToSocketAddrs,
   | |_____________________- first implementation here
...
54 | / impl<T> TryFrom<&(Option<T>, T)> for SocketAddrs
55 | | where
56 | |     T: ToSocketAddrs,
   | |_____________________^ conflicting implementation for `SocketAddrs`
   |
   = note: upstream crates may add a new impl of trait `std::net::ToSocketAddrs` for type `std::option::Option<_>` in future versions
```
Any ideas?